pkg/asset/installconfig/alibabacloud/validation.go

```diff
@@ -15,168 +14,140 @@ func Validate(client *Client, ic *types.InstallConfig) error {
-			continue
-		}
-
-		isWithStock := false
-		availableResources := response.AvailableZones.AvailableZone
-		if availableResources != nil {
-			supportedResources := availableResources[0].AvailableResources.AvailableResource
-			if supportedResources != nil {
-				resource := supportedResources[0].SupportedResources.SupportedResource
-				if resource != nil && resource[0].StatusCategory == "WithStock" {
-					isWithStock = true
-				}
+		for _, zone := range pool.Zones {
+			if !availableZones[zone] {
+				allErrs = append(allErrs, field.Invalid(fldPath, zone, fmt.Sprintf("zone ID is unavailable in region %q", ic.Platform.AlibabaCloud.Region)))
 			}
 		}
-		if !isWithStock {
-			allErrs = append(allErrs, field.Invalid(path, instanceType, fmt.Sprintf("instance type is out of stock in zone %q", zoneID)))
+	}
+	// InstanceType and zones are related.
+	// If the availability zone is not available, the instanceType will not be validated.
+	if len(allErrs) == 0 {
+		if pool.InstanceType != "" {
+			allErrs = append(allErrs, validateInstanceType(client, pool.Zones, pool.InstanceType, fldPath)...)
 		}
 	}
+
 	return allErrs
 }
 
-func validatePlatform(client *Client, ic *types.InstallConfig, path *field.Path) field.ErrorList {
+func validateInstanceType(client *Client, zones []string, instanceType string, path *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	fldpath := path.Child("instanceType")
+	zoneWithStock := map[string]bool{}
 
-	if ic.AlibabaCloud.ResourceGroupID != "" {
-		allErrs = append(allErrs, validateResourceGroup(client, ic, path)...)
+	response, err := client.DescribeAvailableZoneByInstanceType(instanceType)
+	if err != nil {
+		return append(allErrs, field.InternalError(fldpath, err))
 	}
-
-	if ic.Platform.AlibabaCloud.DefaultMachinePlatform != nil {
-		allErrs = append(allErrs, validateMachinePool(client, ic, path.Child("defaultMachinePlatform"), ic.Platform.AlibabaCloud.DefaultMachinePlatform, nil)...)
+	if len(response.AvailableZones.AvailableZone) == 0 {
+		return append(allErrs, field.Invalid(fldpath, instanceType, "no available availability zones found"))
 	}
 
-	return allErrs
-}
-
-func validateMachinePool(client *Client, ic *types.InstallConfig, fldPath *field.Path, pool *alibabacloudtypes.MachinePool, replicas *int64) field.ErrorList {
-	allErrs := field.ErrorList{}
-	var zones []string
-	response, err := client.DescribeAvailableResource("Zone")
-	if err != nil {
-		return append(allErrs, field.InternalError(fldPath, err))
-	}
 	for _, zone := range response.AvailableZones.AvailableZone {
-		if zone.Status == "Available" {
-			zones = append(zones, zone.ZoneId)
+		if zone.Status == "Available" && zone.StatusCategory == "WithStock" {
+			zoneWithStock[zone.ZoneId] = true
 		}
 	}
+	if len(zoneWithStock) == 0 {
+		return append(allErrs, field.Invalid(fldpath, instanceType, "no availability zone with stock found"))
+	}
```

Comment:
This should only be checked if `zones` is empty. Otherwise, we want to create an error for each zone explicitly specified by the user.